firmware/Core/Src/gps/gps_internal_drivers.c

```diff
@@ -0,0 +1,144 @@
+		UART_gps_port_handle,
+		(uint8_t *)cmd_buf,
+		cmd_buf_len,
+		GPS_RX_TIMEOUT_MS); // FIXME: update the timeout
+
+	if (tx_status != HAL_OK)
+	{
+		LOG_message(
+			LOG_SYSTEM_GPS, LOG_SEVERITY_WARNING, LOG_SINK_ALL,
+			"GPS ERROR: tx_status != HAL_OK (%d)\n",
+			tx_status);
+
+		return 1;
+	}
+
+	// FIXME: Update the timeouts with the actual times, it currently works with 800 ms and 500ms does not work
+	// GPS takes time to respond, first section of log response ie <OK\n [COM1] is quick but the rest of the data response takes a while
+	
+
+	// RX FROM GPS, into UART_gps_buffer
+	const uint32_t start_rx_time = HAL_GetTick();
+	while (1) {
+		if ((UART_gps_buffer_write_idx == 0)) {
+			// Check if we've timed out (before the first byte)
+			if ((HAL_GetTick() - start_rx_time) > GPS_RX_TIMEOUT_MS) {
+
+				LOG_message(
+					LOG_SYSTEM_GPS, LOG_SEVERITY_WARNING, LOG_SINK_ALL,
+					"GPS ERROR: Timeout before receiving any data"
+				);
+			
+				// fatal error; return
+				return 2;
+			}
+		}
+		else { // thus, UART_eps_buffer_write_idx > 0
+			// Check if we've timed out (between bytes)
+			const uint32_t cur_time = HAL_GetTick();
+			// Note: Sometimes, because ISRs and C are fun, the UART_eps_last_write_time_ms is
+			// greater than `cur_time`. Thus, we must do a safety check that the time difference
+			// is positive.
+			if (
+				(cur_time > UART_gps_last_write_time_ms) // Important seemingly-obvious safety check.
+				&& ((cur_time - UART_gps_last_write_time_ms) >  GPS_RX_TIMEOUT_MS)
+			) {
+				LOG_message(
+					LOG_SYSTEM_GPS, LOG_SEVERITY_WARNING, LOG_SINK_ALL,
+					"GPS WARNING: Timeout during receiving of response. UART_gps_last_write_time_ms=%lu, cur_time=%lu",
+					UART_gps_last_write_time_ms,
+					cur_time
+				);
+				// Non-fatal error; try to parse what we received
+				break;
+			}
+		}
+	}
+
+	// End Receiving
+    // TODO: Determine where to call this. Further testing required
+    // If nothing is being populated in the UART_gps_buffer during testing, consider commenting out the line below
+	GPS_set_uart_interrupt_state(0); // We are no longer expecting a response
+
+	// Logging the received response
+	LOG_message(
+		LOG_SYSTEM_GPS, LOG_SEVERITY_NORMAL, LOG_SINK_ALL,
+        "GPS Buffer Data: %s",
+        UART_gps_buffer
+	);
+
+	// Check that we've received what we're expecting
+	// TODO: if the following cases happen ever during testing, consider allowing them and treating them as WARNINGs
+	if (UART_gps_buffer_write_idx == 0)
+	{
+		LOG_message(
+			LOG_SYSTEM_GPS, LOG_SEVERITY_WARNING, LOG_SINK_ALL,
+			"GPS ERROR: UART_gps_buffer_write_idx == 0"
+		);
+
+		return 3;
+	}
+	else if (UART_gps_buffer_write_idx > rx_buf_max_size)
+	{
+		LOG_message(
+			LOG_SYSTEM_GPS, LOG_SEVERITY_WARNING, LOG_SINK_ALL,
+			"GPS ERROR: UART_gps_buffer overflow");
+		return 4;
+	}
+
+	// Copying the log response from the UART gps buffer to the rx_buf[] and clearing the buffer
```

Comment:
Change "Copying" to "Copy". Action comments are generally written in imperative tense

Reply:
Done